### PR TITLE
Enabling user in contentChangeDetails

### PR DESCRIPTION
**Pull request by shtukas**
This change brings the setting of the user in the contentChangeDetails of atoms.
It builds on the work made in the previous change :)

### CODE REVIEW

File: explainer-server/app/models/ExplainerStore.scala [hunk 1]
```diff
@@ -18,17 +18,19 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
 
   val explainerDB = new ExplainerDB(config)
 
-  def buildAtomWithDefaults(id: String, explainerAtom: ExplainerAtom, revision: Long): Atom = {
+  def buildAtomWithDefaults(user: com.gu.pandomainauth.model.User, id: String, explainerAtom: ExplainerAtom, revision: Long): Atom = {
```
Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala]: import the class to clean this up and maybe move to end of parameter list

Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala]: take contentChangeDetails as parameter

File: explainer-server/app/models/ExplainerStore.scala [hunk 2]
```diff
@@ -46,16 +48,16 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
...
   }
 
-  def create(): Future[Atom] = {
+  def create(user: com.gu.pandomainauth.model.User): Future[Atom] = {
     val uuid = java.util.UUID.randomUUID.toString
     val explainerAtom = ExplainerAtom("-", "-", DisplayType.Expandable)
-    val explainer = buildAtomWithDefaults(uuid, explainerAtom, 1)
+    val explainer = buildAtomWithDefaults(user, uuid, explainerAtom, 1)
     explainerDB.store(explainer)
     Future(explainer)
```
Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala:82]: Add a publish function here which updates the published field of contentChangeDetails

File: explainer-server/app/models/ExplainerStore.scala (hunk 1 above)
Comment (shtukas) [inline explainer-server/app/models/ExplainerStore.scala]: Both done!

File: explainer-server/app/models/ExplainerStore.scala [hunk 3]
```diff
@@ -46,18 +48,37 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
...
       updatedExplainer
     }
   }
 
-  def create(): Future[Atom] = {
+  def create(user: PandaUser): Future[Atom] = {
     val uuid = java.util.UUID.randomUUID.toString
     val explainerAtom = ExplainerAtom("-", "-", DisplayType.Expandable)
-    val explainer = buildAtomWithDefaults(uuid, explainerAtom, 1)
+    val contentChangeDetails = ContentChangeDetails(
```
Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala]: Add created ChangeRecord here

File: explainer-server/app/models/ExplainerStore.scala [hunk 4]
```diff
@@ -14,21 +14,41 @@ import scala.concurrent.Future
...
+    def buildChangeRecord(existingRecord: Option[ChangeRecord], shouldUpdate: Boolean) = {
+      if (shouldUpdate) {
+        Some(ChangeRecord(DateTime.now.getMillis, user=pandaUserToAtomUser(user)))
+      } else if (existingRecord.isDefined) existingRecord else None
+    }
+    ContentChangeDetails(
+      created      = buildChangeRecord(existingContentChangeDetails.flatMap(_.created)     , updateCreated),
+      lastModified = buildChangeRecord(existingContentChangeDetails.flatMap(_.lastModified), updateLastModified),
+      published    = buildChangeRecord(existingContentChangeDetails.flatMap(_.published)   , updatePublished),
+      revision     = existingContentChangeDetails.map(_.revision).getOrElse(1)
```
Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala]: when updating, we need to update this. maybe 

`existingContentChangeDetails.map(_.revision).getOrElse(0)+1`

File: explainer-server/app/models/ExplainerStore.scala [hunk 5]
```diff
@@ -46,18 +66,31 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
...
+    val explainer = buildAtomWithDefaults(uuid, explainerAtom, contentChangeDetails, user)
     explainerDB.store(explainer)
     Future(explainer)
   }
+
+  def publish(id: String, user: PandaUser): Future[Atom] = {
+    explainerDB.load(id).map{ explainer =>
+      val newExplainerAtom = ExplainerAtom(explainer.tdata.title, explainer.tdata.body, explainer.tdata.displayType)
+      val contentChangeDetails = contentChangeDetailsBuilder(user, explainer.contentChangeDetails, false, false, true)
+      val updatedExplainer = buildAtomWithDefaults(explainer.id, newExplainerAtom, contentChangeDetails, user)
```
Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala]: newExplainerAtom is unnecessary here, you can just user explainer.data

File: explainer-server/app/models/ExplainerStore.scala [hunk 6]
```diff
@@ -46,18 +66,31 @@ class ExplainerStore @Inject() (config: Config) extends ExplainerAtomImplicits
...
-    val explainer = buildAtomWithDefaults(uuid, explainerAtom, 1)
+    val contentChangeDetails = contentChangeDetailsBuilder(user, None, true, true, false)
+    val explainer = buildAtomWithDefaults(uuid, explainerAtom, contentChangeDetails, user)
     explainerDB.store(explainer)
     Future(explainer)
   }
+
+  def publish(id: String, user: PandaUser): Future[Atom] = {
+    explainerDB.load(id).map{ explainer =>
+      val newExplainerAtom = ExplainerAtom(explainer.tdata.title, explainer.tdata.body, explainer.tdata.displayType)
```
Comment (philmcmahon) [inline explainer-server/app/models/ExplainerStore.scala]: this shouldn't be necessary - just use explainer.data

**Comment by philmcmahon** [comment]
👍

**Pull request merged**